vpr/src/place/initial_placement.cpp

```diff
@@ -0,0 +1,426 @@
+                // go through the chip exhaustively to find a legal placement for the macro
+                // place the macro on the first location that is legal
+                // then set macro_placed = true;
+                // if there are no legal positions, error out
+
+                // Exhaustive placement of carry macros
+                for (ipos = 0; ipos < free_locations[itype] && macro_placed == false; ipos++) {
+                    // Try to place the macro
+                    macro_placed = try_place_macro(itype, ipos, pl_macro);
+
+                } // Exhausted all the legal placement position for this macro
+
+                // If macro could not be placed after exhaustive placement, error out
+                if (macro_placed == false) {
+                    // Error out
+                    VPR_FATAL_ERROR(VPR_ERROR_PLACE,
+                                    "Initial placement failed.\n"
+                                    "Could not place macro length %zu with head block %s (#%zu); not enough free locations of type %s (#%d).\n"
+                                    "Please manually size the FPGA because VPR can't do this yet.\n",
+                                    pl_macro.members.size(), cluster_ctx.clb_nlist.block_name(blk_id).c_str(), size_t(blk_id), device_ctx.physical_tile_types[itype].name, itype);
+                }
+
+            } else {
+                // This macro has been placed successfully, proceed to place the next macro
+                continue;
+            }
+        }
+    } // Finish placing all the pl_macros successfully
+}
+
+/* Place blocks that are NOT a part of any macro.
+ * We'll randomly place each block in the clustered netlist, one by one. */
+static void initial_placement_blocks(int* free_locations, enum e_pad_loc_type pad_loc_type) {
+    int itype, ipos;
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+
+    // The map serves to place first the most constrained block ids
+    std::map<int, std::vector<ClusterBlockId>> sorted_block_map;
+
+    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+        auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
+
+        size_t num_equivalent_tiles = logical_block->equivalent_tiles.size();
+        sorted_block_map[num_equivalent_tiles].push_back(blk_id);
+    }
+
+    for (auto& sorted_blocks : sorted_block_map) {
+        for (auto blk_id : sorted_blocks.second) {
+            if (place_ctx.block_locs[blk_id].loc.x != -1) { // -1 is a sentinel for an empty block
+                // block placed.
+                continue;
+            }
+
+            auto logical_block = cluster_ctx.clb_nlist.block_type(blk_id);
+
+            /* Don't do IOs if the user specifies IOs; we'll read those locations later. */
+            if (!(is_io_type(pick_random_physical_type(logical_block)) && pad_loc_type == USER)) {
```

Comment:
Both here and below, we probably shouldn't be using 'pick_random_physical_type()', since that may return a different type each time. We should pick a single type to try and be consistent through-out.

Otherwise a different random choice of type may be made which would be inconsistent (may pass a check one time, but not the same check again later).

Reply:
Done